org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ReferencesHandler.java

```diff
@@ -53,13 +56,16 @@ public ReferencesHandler(PreferenceManager preferenceManager) {
 		this.preferenceManager = preferenceManager;
 	}
 
-	private IJavaSearchScope createSearchScope() throws JavaModelException {
+	private IJavaSearchScope createSearchScope(IJavaElement elementToSearch) throws JavaModelException {
 		IJavaProject[] projects = JavaCore.create(ResourcesPlugin.getWorkspace().getRoot()).getJavaProjects();
```

Comment:
Its worth to look into JDT https://github.dev/eclipse-jdt/eclipse.jdt.core/blob/a24736c1a409b3b46c0e5a29cbc29d7b672afa9f/org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/BasicSearchEngine.java#L141

Or try to use the same scope creation method. Since vscode has no ability to provide in which scopes we should search on, it better to include all scopes.

Should we only include Application Libraries only if the client support class file content ? which I assume if there is a decompiler configured. If I'm not mistaken the JDT Core is cable of indexing bytecode without source. So basically the search engine can provide the results, But instead of showing the source we could who the skeleton so that end user knows that they need a decompiler. WDYT ?

Reply:
I haven't changed it.

Reply:
May be i was not clear, When looking at your implementation I see a difference how JDT and JDT.LS will create the search scope. For example `IJavaSearchScope.SYSTEM_LIBRARIES` do we really want to add it only if the element is inside JRE ? The main question is in JDT `IClasspathContainer.K_SYSTEM` is only used for JRE, because searching for `IJavaSearchScope.SYSTEM_LIBRARIES` represents searching in containers which are either `IClasspathContainer.K_SYSTEM` or `IClasspathContainer.K_DEFAULT_SYSTEM`. Do you see otherwise ?

Even though it is not directed connected to this PR, I also see `IJavaSearchScope.APPLICATION_LIBRARIES` is added only if we have a support for class file content, which means a decompiler configured. I assume this works even when JDT.LS inbuilt decompiled source skeleton is used as well.

Reply:
I have updated the PR.

Reply:
Still do you think with the current approach we will find reference of a class in a classpath entry which is marked as K_SYSTEM ? I think search scope is a way of narrowing down where we would like to search in Eclipse JDT UI. For example you can

- Search only in main source
- Include dependent projects
- Include the runtime like JRE or anyother classpaths that are marked as such
- Search only in test source
- Workingsets
etc

So this mainly filter out unwanted results by not searching on such classpath entries, basically the scope narrow down the number indexes we search on.

But in vscode we have no such feature to filter where we would like to search the references. One thing I could see is that, Goto Reference only search on source and test, Find All References search every where. But thats a different discussion I think since the LSP doesn't have two protocol messages for that.

So my idea is that we fix this issue by including all scopes for now. Let me know what you think.

Reply:
I have copied isInsideJRE from jdt.ui - https://github.com/eclipse/eclipse.jdt.ls/pull/2413/commits/c3cac4f32d84e2f36fc6e3bcb13bcb9b07d976b8#diff-168bf4012038f5d2d21ac8e4639f5e6d4a10c28cc99abb48588ea318352dc4cbR154